mdn/developer-portal
# 1438: Add new aspect ratio card image

This changeset follows the clean-up work in PR #1454  and adds support for a new "card image", this this time in a 3:2 aspect ratio rather than the new default of 16:9.

If the 3:2 image is not available/not configured, a 3:2 fallback will be used.

(Related issue #1438 )

## Key changes:

-  Add support for a 3:2-aspect-ratio card image on certain Page types

    Specifically those which can appear as 'Featured' items, where certain configurations can result in 3:2 aspect-ratio images being used instead of the new default of 16:9:
    * `Article`s
    * `Video`s
    * `Event`s
    * `ExternalContent` (ie, `ExternalArticle`s, `ExternalVideo`s, `ExternalEvent`s)

    Note that `ContentPage`, `Topic` and `Person` Page types do NOT get the 3:2 field added, because we don't need it yet and it's confusing to add it if we don't use it

-  Add support for a 3:2 image in all our custom StreamBlocks
    The 3:2 image is optional, just to make life slightly easier for cases where we know we don't need to worry about it being rendered as a 3:2 image.

- Complete support of 3:2 images for featured items by picking the appropriate image in the HTML

    Includes:
    * Pages (ie, using the appropriate "card image")
    * StreamBlocks (ie, using the appropriate "image" field)
    * Fallback images if either of the above is needed but not present

-  FIX: Ensure the bottom row of featured items in a two-rows-of-two layout uses a 3:2 aspect ratio image, too

## Screenshots

**Example of a page now featuring both 16:9 and 3:2 image options**
![Screenshot 2020-05-11 at 11 52 21](https://user-images.githubusercontent.com/101457/81565498-5727f980-9391-11ea-84a8-b4f7bfcbc9bf.png)

**Example of StreamField block (ie ad-hoc content, not a page) now featuring both 16:9 and 3:2 image options** -- this is useful for when we want to feature an ad-hoc external link on the homepage, for instance - but not that the importer will not automatically import a 3:2 image, so that will need manual handling as/when required.

![Screenshot 2020-05-11 at 13 08 44](https://user-images.githubusercontent.com/101457/81565501-58592680-9391-11ea-9c27-8af11db1f1a4.png)

**Example of a featured external link switching between 3:2 and 16:9 images** depending on where it is configured. (ie, this is a `StreamField`'s `StreamBlock` showing the behaviour, but a featured Page will also behave the same way.
![Screenshot 2020-05-11 at 13 13 22](https://user-images.githubusercontent.com/101457/81565502-598a5380-9391-11ea-8d04-00c3174268d4.png)
![Screenshot 2020-05-11 at 13 13 43](https://user-images.githubusercontent.com/101457/81565510-5becad80-9391-11ea-8079-9646c94777a4.png)

**And an example of a 3:2 card lacking a 3:2 image falling back to the correct-aspect-ratio placeholder**
![Screenshot 2020-05-11 at 13 22 04](https://user-images.githubusercontent.com/101457/81565511-5d1dda80-9391-11ea-909d-8251c748872a.png)

## How to test

- Code is up on dev. I've also added a test 3:2 and test 16:9 image. Please drop these into various pages and featured-elements configurations (including re-ordering them) and previewing/publishing the pages to see different behaviour. Please try to break it!
